test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -833,6 +833,66 @@ UCS_TEST_P(test_ucp_sockaddr, compare_cm_and_wireup_configs) {
 
 UCP_INSTANTIATE_ALL_TEST_CASE(test_ucp_sockaddr)
 
+
+class test_ucp_sockaddr_different_tl_rsc : public test_ucp_sockaddr {
+protected:
+    void init() {
+        m_err_count = 0;
+        get_sockaddr();
+        test_base::init();
+
+        /* entities will be created in a test */
+    }
+
+    void cleanup() {
+        /* cleanup will be done in a test */
+    }
+
+    // create entities with different set of MDs and TL resources on a client
+    // and on a server to test non-homogeneous setups
+    void do_test(bool self_devices, bool shm_devices) {
+        if (self_devices) {
+            modify_config("SELF_DEVICES", "");
+        }
+        if (shm_devices) {
+            modify_config("SHM_DEVICES", "");
+        }
+
+        // create a client with restrictions
+        create_entity();
+
+        if (self_devices) {
+            modify_config("SELF_DEVICES", "all");
+        }
+        if (shm_devices) {
+            modify_config("SHM_DEVICES", "all");
+        }
+
+        // create a server with restrictions
+        if (!is_self()) {
+            create_entity();
+        }
+
+        skip_loopback();
+        listen_and_communicate(false, SEND_DIRECTION_BIDI);
+        test_ucp_sockaddr::cleanup();
+    }
+};
+
+
+UCS_TEST_P(test_ucp_sockaddr_different_tl_rsc, unset_devices_and_communicate) {
+    bool flags[] = { false, true };
+
+    for (uint8_t self_idx = 0; self_idx < 2; ++self_idx) {
+        for (uint8_t shm_idx = 0; shm_idx < 2; ++shm_idx) {
+            do_test(flags[self_idx], flags[shm_idx]);
+        }
+    }
+}
```

Comment:
could these be different test variants?
this way, can avoid the direct call to `test_base::init();` and create entities in ucp_test::init as usual

Reply:
done
but anyway can't avoid `init()`